pkg/ccl/backupccl/restore_span_covering.go

```diff
@@ -78,20 +99,55 @@ func makeSimpleImportSpans(
 
 		for layer := range backups {
 			covPos := spanCoverStart
+			var curCovGrowSize int64
+
 			// TODO(dt): binary search to the first file in required span?
 			for _, f := range backups[layer].Files {
 				if sp := span.Intersect(f.Span); sp.Valid() {
 					fileSpec := execinfrapb.RestoreFileSpec{Path: f.Path, Dir: backups[layer].Dir}
 					if dir, ok := backupLocalityMap[layer][f.LocalityKV]; ok {
 						fileSpec = execinfrapb.RestoreFileSpec{Path: f.Path, Dir: dir}
 					}
+
+					// Lookup the size of the file being added; if the backup didn't
+					// record a file size, just assume it is 16mb for estimating.
+					sz := f.EntryCounts.DataSize
+					if sz == 0 {
+						sz = 16 << 20
+					}
+
 					if len(cover) == spanCoverStart {
 						cover = append(cover, makeEntry(span.Key, sp.EndKey, fileSpec))
+						curCovGrowSize = sz
 					} else {
-						// Add each file to every matching partition in the cover.
+						// If this file extends beyond the end of the last partition of the
+						// cover, either append a new partition for the uncovered span or
+						// grow the last one if size allows.
+						if covEnd := cover[len(cover)-1].Span.EndKey; sp.EndKey.Compare(covEnd) > 0 {
+							// If adding the item size to the current rightmost span size will
+							// exceed the target size, make a new span, otherwise extend the
+							// rightmost span to include the item.
+							if curCovGrowSize+sz > targetSize {
+								cover = append(cover, makeEntry(covEnd, sp.EndKey, fileSpec))
+								curCovGrowSize = sz
+							} else {
+								cover[len(cover)-1].Span.EndKey = sp.EndKey
+								cover[len(cover)-1].Files = append(cover[len(cover)-1].Files, fileSpec)
+								curCovGrowSize += sz
+							}
+						}
+						// Now ensure the file is included in any partition in the existing
+						// cover which overlaps.
 						for i := covPos; i < len(cover) && cover[i].Span.Key.Compare(sp.EndKey) < 0; i++ {
 							if cover[i].Span.Overlaps(sp) {
-								cover[i].Files = append(cover[i].Files, fileSpec)
+								// This file needs to be in this partition; if it it was just
+								// added above it'll be on the end, otherwise add it.
+								if l := len(cover[i].Files) - 1; l < 0 || cover[i].Files[l] != fileSpec {
+									cover[i].Files = append(cover[i].Files, fileSpec)
+									if i == l {
+										curCovGrowSize += sz
```

Comment:
I might be missing something here. why do we need grow `curCovGrowSize` here for when `i==l`? 

I thought this loop is growing "older" covers (i.e. not the current, rightmost cover), potentially beyond the targetSize limit.  I'm under the impression that curCovGrowSize tracks the right most cover's size, so I don't follow why it gets incremented here.

Perhaps this logic could use another comment. It's quite hard to follow all the indexing!

Reply:
`i == l` is the rightmost

Reply:
didn't we already grow the right most span above?

Reply:
aha! This makes sense now. thanks for clarifying!